File: PushLanguageInterpreter/PushLanguageInterpreter/Parser.cs
using PushLanguageInterpreter.Nodes;

namespace PushLanguageInterpreter;

public class Parser
{
    private int _position;

    public IReadOnlyCollection<INode> ParseProgram(IReadOnlyList<Token> program)
    {
        var result = new List<INode>();
        while (_position < program.Count && (program[_position].Value == "(" || char.IsLetter(program[_position].Value[0])))
        {
            if (program[_position].Value == "(")
            {
                _position++;
                var instructions = ParseProgram(program);
                if (program[_position].Value != ")")
                {
                    throw new Exception("Expected ')', got " + program[_position].Value);
                }

                result.Add(new ProgramNode(instructions));
            }
            else if (program[_position].Value.Contains('_'))
            {
                result.Add(ParseInstruction(program[_position]));
            }
            else
            {
                result.Add(new FloatNode(double.Parse(program[_position].Value)));
            }

            _position++;
        }

        return result;
    }

    private static InstructionNode ParseInstruction(Token token)
    {
        var parts = token.Value.Split('_');
        return new InstructionNode(Enum.Parse<NodeType>(parts[0], true), parts[1]);
    }
}

File: genetic-programming-scala-sandbox/untitled/src/main/java/com/yaskovdev/gp/push/Parser.java
package com.yaskovdev.gp.push;

import com.yaskovdev.gp.push.node.FloatNode;
import com.yaskovdev.gp.push.node.InstructionNode;
import com.yaskovdev.gp.push.node.Node;
import com.yaskovdev.gp.push.node.ProgramNode;

import java.util.ArrayList;
import java.util.List;

import static java.lang.Double.parseDouble;

public class Parser {

    private int position;

    public List<Node> parseProgram(List<Token> program) {
        final List<Node> result = new ArrayList<>();
        while (position < program.size() && program.get(position).isNotCloseBracket()) {
            if (program.get(position).isOpenBracket()) {
                position++;
                final List<Node> instructions = parseProgram(program);
                if (program.get(position).isNotCloseBracket()) {
                    throw new RuntimeException("Expected ')', got " + program.get(position));
                }
                result.add(new ProgramNode(instructions));
            } else if (program.get(position).isInstruction()) {
                result.add(new InstructionNode(program.get(position).getValue()));
            } else {
                result.add(new FloatNode(parseDouble(program.get(position).getValue())));
            }
            position++;
        }
        return result;
    }
}
